docs: add concepts guide on Datatypes and Datashapes (#8557)
ibis-project/ibis: docs/concepts/datatypes_datashapes.qmd

> # Datatypes and Datashapes

Reviewer: ```suggestion
---
title: Datatypes and Datashapes
---
```

for multi-word titles, this also handles setting the page name in a browser tab

Every value in Ibis has two important properties: a type and shape.

The type is probably familiar to you. It is something like

- `Integer`
- `Floating`
- `String`
- `Array`

The shape is one of

- `Scalar` (a single value)
- `Column` (a series of values)

## Datatype Flavors

For some datatypes, there are further options that define them.
For instance, `Integer` values can be signed or unsigned, and
they have a precision. For example, "uint8", "int64", etc.
These flavors don't affect their capabilities
(eg both signed and unsigned ints have a `.abs()` method),
but the flavor does impact how the underlying backend performs the computation.

## Capabilities

Depending on the combination of datatype and datashape, a value has
different capabilities. For example:

- All `String` values (both `StringScalars` and `StringColumns`) have the
  method `.upper()` that transforms the string to uppercase.
  `Floating` and `Array` values don't have this method, of course.
- `IntegerColumn` and `FloatingColumn`  values have `.mean()`, `.max()`, etc methods
  (which return `FloatingScalar` values),
  because you can aggregate over them, since they are a collection of values.

>   On the other hand, `IntegerScalar` and `FloatingColumn` values do **not** have these

Reviewer: ```suggestion
  On the other hand, `IntegerScalar` and `FloatingScalar` values do **not** have these
```

  methods, because it doesn't make sense to take the mean or max of a single value.

> - If you call `.execute()` on these values, you get different results.

Reviewer: ```suggestion
- If you call `.to_pandas()` on these values, you get different results.
```

or `to_pyarrow`?  but we're trying to get away from `execute`

Reviewer: ahh, I see below you are referencing pandas datashapes, so definitely `to_pandas`, then

  `Scalar` shapes result in plain python objects:
  - `IntegerScalar`: plain python `int` object.
  - `FloatingScalar`: plain python `float` object.
  - `StringScalar`: plain python `str` object.
  - `ArrayScalar`: plain python `list` object.
  On the other hand, `Column` shapes result in `pandas.Series`:
  - `IntegerColumn`: pd.Series of integers, with the same flavor.
       For example, if the `IntegerColumn` was specifically "uint16",
       then the pandas series will also be of type "uint16".
  - `FloatingColumn`: pd.Series of floats with the same flavor.
  - etc.

## Broadcasting and Alignment

There are rules for how different datashapes are combined. This is similar to

> how SQL and numpy handles merging datashapes, if you are familiar with them.

Reviewer: ```suggestion
how SQL and NumPy handles merging datashapes, if you are familiar with them.
```

I think it's always NumPy in prose

```python
import ibis

> ibis.options.interactive = True

Reviewer: Maybe make this whole section executable to handle rendering the output?

# │            3007 │
# │            3007 │
# │            3007 │
# │            3007 │
# │            3007 │
# │            3007 │
# │            3007 │
# │            3007 │
# │            3007 │
# │            3007 │
# │               … │
# └─────────────────┘

# Combining two Columns results in a Column:
t1.year + t1.bill_length_mm
# ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
# ┃ Add(year, bill_length_mm) ┃
# ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
# │ float64                   │
# ├───────────────────────────┤
# │                    2046.1 │
# │                    2046.5 │
# │                    2047.3 │
# │                      NULL │
# │                    2043.7 │
# │                    2046.3 │
# │                    2045.9 │
# │                    2046.2 │
# │                    2041.1 │
# │                    2049.0 │
# │                         … │
# └───────────────────────────┘
```

One requirement that might surprise you if you are coming from Numpy is
Ibis's requirements on aligning `Columns`: In Numpy, if you have two arbitrary
arrays, each of length 100, you can add them together, and it works because the
elements are "lined up" based on position. Ibis is different. Because it is based

> around SQL, and SQL has no notion of inherint row ordering, you cannot "line up"

Reviewer: ```suggestion
around SQL, and SQL has no notion of inherent row ordering, you cannot "line up"
```
